api/src/test/java/org/pmiops/workbench/api/AdminAuthorityTest.java

```diff
@@ -0,0 +1,197 @@
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import org.junit.jupiter.api.Test;
+import org.pmiops.workbench.annotations.AuthorityRequired;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+public class AdminAuthorityTest {
+
+  /**
+   * API endpoints which are actually admin endpoints, but don't conform to the /admin/ infix
+   * convention. Ideally these REST paths should eventually be updated for consistency. Avoid adding
+   * more entries to this list and put /admin/ in the REST path instead.
+   */
+  private static final Set<String> NONCONFORMING_ADMIN_METHODS =
+      ImmutableSet.of(
+          "postStatusAlert",
+          "publishWorkspace",
+          "unpublishWorkspace",
+          "getInstitution",
+          "getInstitutions",
+          "createInstitution",
+          "deleteInstitution",
+          "updateInstitution",
+          "setInstitutionUserInstructions",
+          "deleteInstitutionUserInstructions",
+          "createAuthDomain",
+          "updateUserDisabledStatus");
+
+  /**
+   * Conversely, methods which have an "/admin/" infix, but do not require an Authority. This should
+   * be very exceptional, and should entail some additional method of authentication.
+   *
+   * <p>Please add a comment for any exceptions added to this list.
+   */
+  private static final Set<String> NONAUTHORITY_ADMIN_METHODS =
+      ImmutableSet.of(
+          // logEgressEvent requires an API key secret; it's a webhook which cannot use authority
+          "logEgressEvent",
+          // Not technically an admin endpoint, but only enabled in lower environments
+          "unsafeSelfBypassAccessRequirement");
+
+  /**
+   * Heuristic test: if there's a REST endpoint that looks like an admin endpoint (based on the
+   * URL), ensure that it has an @AuthorityRequired annotation. Assert the inverse as well.
+   * Exceptions are enumerated in the above allowlists.
+   *
+   * <p>To achieve this, we use reflection over our API controller methods as well as the Swagger
+   * generated interfaces. There are three relevant classes:
+   *
+   * <ul>
+   *   <li>our source code controller implementation, which may or may not have @AuthorityRequired
+   *   <li>the generated FooApiDelegate interface, which our controller implements
+   *   <li>and the generated FooApi interface, which integrates with Spring, and contains an
+   *       annotation @RequestMapping with request path information
+   * </ul>
+   *
+   * <p>We walk these relationships to check our heuristic. The following preconditions must hold:
+   *
+   * <ul>
+   *   <li>all API controllers live in the RW "api" Java package
```

Comment:
and have the annotation RestController

Reply:
Added. Also added another assumption about the codegen to the list